_posts/2024-01-23-ensuring-code-quality-a-guide-to-dynamic-git-hub-pull-request-gates.markdown

```diff
@@ -0,0 +1,309 @@
+    uses: ./.github/workflows/test.execution.yml
+    secrets: inherit
+    with:
+      app-url: ${{ '{{' }} needs.build-deploy.outputs.app-url }}
+```
+
+The calling job `test` can't be set as status check because GitHub ignores its status.
+
+Configuring a job in the called reusable workflow `test.execution.yml` is not an option either.<br>
+If `test` in the calling workflow would be skipped, the pull request would wait forever for the status check in the 
+reusable workflow as the skipping happens on workflow parent level, not child level.
+
+The workaround is to configure an additional, standard job `test-completed` that concludes the workflow as status 
+check and fails if the `test` failed before.<br>
+Setting `if: always()` ensures that the job is always executed, even if the `test` job failed.
+
+```yaml
+on:
+  pull_request:
+    
+jobs:
+  check-if-relevant:
+    runs-on: ubuntu-latest
+    outputs:
+      is-relevant: ${{ '{{' }} steps.relevant-files-changed.outputs.any_changed }}
+    steps:
+     [...]
+ 
+  build-deploy:
+    needs: check-if-relevant
+    if: needs.check-if-relevant.outputs.is-relevant == 'true'
+    uses: ./.github/workflows/release.build_deploy.yml
+    secrets: inherit  
+
+  test:
+    needs: build-deploy
+    uses: ./.github/workflows/test.execution.yml
+    secrets: inherit
+    with:
+      app-url: ${{ '{{' }} needs.build-deploy.outputs.app-url }}
+
+  test-completed:
+    needs: test
+    runs-on: 'ubuntu-latest'
+    if: always()
+    steps:
+      - name: Check test job status
+        if: needs.test.result == 'failure'
+        run: exit 1
+```
+
+## Alternative Approach
+
+Another workaround is to add job like the `required-status-check` in the following example to all dynamically required workflows.
+
+```yaml
+on:
+  pull_request:
+    paths:
+      - src/**
+      - tests/**
+
+jobs:  
+  build-deploy:
+    runs-on: ubuntu-latest
+    steps:
+      [...]
+
+  test:
+    needs: build-deploy
+    runs-on: ubuntu-latest
+    steps:
+      [...]
+    
+  required-status-check:
+    needs: test
+    runs-on: 'ubuntu-latest'
+    if: always()
+    steps:
+      - name: Check workflow status
+        if: needs.test.result != 'success'
+        run: exit 1
+```
+
+This job would return a failure state if the tests of job `test` failed.
+
+In the repo settings configure `required-status-check` as required status check in the branch protection
+rules of the main branch.<br>
+The pull requests get blocked if none of the triggered workflows have implemented the `required-status-check` job or at
+least one of them is failing.<br>
+It gets unblocked if at least one `required-status-check` job is implemented and all of them are successful. 
+And usually, at least one workflow needs to run in any case (e.g. a lint or source formatting check) and can implement
+this.
+
+### Advantages
+
+*   Only those workflows get executed that are actually required, because the list of relevant files are declared at the
```

Comment:
```suggestion
*   Only those workflows get executed that are actually apply to the changed files, because the list of relevant files are declared at the
```